File: presences/src/main/java/fr/openent/presences/service/EventService.java
package fr.openent.presences.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.entcore.common.user.UserInfos;

import java.util.List;

public interface EventService {

    /**
     * Get events
     *
     * @param structureId   structure identifier
     * @param startDate     startDate start date
     * @param endDate       endDate end date
     * @param eventType     event type
     * @param listReasonIds reasonId reason_id
     * @param noReason      noReason filter
     * @param userId        userId userId neo4j
     * @param regularized   regularized filter
     * @param followed      followed filter
     * @param page          page
     * @param handler       function handler returning data
     */

    void getEvents(String structureId, String startDate, String endDate,
                   List<String> eventType, List<String> listReasonIds, Boolean noReason, Boolean noReasonLateness, List<String> userId, JsonArray userIdFromClasses,
                   Boolean regularized, Boolean followed, Integer page, Handler<Either<String, JsonArray>> handler);

    void get(String structureId, String startDate, String endDate, String startTime, String endTime,
             List<String> eventType, List<String> listReasonIds, Boolean noAbsenceReason, Boolean noLatenessReason, List<String> userId, List<String> restrictedClasses,
             Boolean regularized, Boolean followed, Integer page, Handler<AsyncResult<JsonArray>> handler);

    Future<JsonArray> getEventsBetweenDates(String startDate, String endDate, List<String> users, List<Integer> eventType,
                                            String structureId);

    void getDayMainEvents(String structureId, String startDate, String endDate, String startTime, String endTime,
                          List<String> studentIds, List<String> typeIds, List<String> reasonIds, Boolean noAbsenceReason, Boolean noLateness,
                          Boolean regularized, Boolean followed, Integer page, Handler<AsyncResult<JsonArray>> handler);

    /**
     * Get events. Non paginated version. Used to display summary page
     *
     * @param startDate Start date
     * @param endDate   End date
     * @param eventType Event type list
     * @param users     User list
     * @param handler   Function handler returning data
     */
    void get(String startDate, String endDate, List<Number> eventType, List<String> users, Handler<Either<String, JsonArray>> handler);

    /**
     * Get events page number
     *
     * @param structureId      structure identifier
     * @param startDate        startDate start date
     * @param endDate          endDate end date
     * @param eventType        event type
     * @param listReasonIds    reasonId reason_id
     * @param noAbsenceReason  noAbsenceReason filter
     * @param noLatenessReason noLatenessReason filter
     * @param userId           userId userId neo4j
     * @param regularized      filter regularized absence
     * @param followed         filter followed absence
     * @param handler          function handler returning data
     */
    void getPageNumber(String structureId, String startDate, String endDate, String startTime, String endTime,
                       List<String> eventType, List<String> listReasonIds, Boolean noAbsenceReason, Boolean noLatenessReason, List<String> userId,
                       Boolean regularized, Boolean followed, Handler<Either<String, JsonObject>> handler);


    /**
     * Get absence and absence events counts summary.
     *
     * @param structureId     structure identifier
     * @param startAt         start date
     * @param endAt           end date
     * @param requiredMarkers required makers
     * @return future with results
     */
    Future<JsonObject> getAbsencesCountSummary(String structureId, String startAt, String endAt, List<String> requiredMarkers);

    Future<JsonObject> getAbsencesCountSummary(String structureId, String startAt, String endAt);


    /**
     * Create event
     *
     * @param event   event
     * @param user    user that create event
     * @param handler function handler returning data
     */
    void create(JsonObject event, UserInfos user, Handler<Either<String, JsonObject>> handler);

    /**
     * Update given event
     *
     * @param id      event identifier
     * @param event   event
     * @param handler Function handler returning data
     */
    void update(Integer id, JsonObject event, Handler<Either<String, JsonObject>> handler);

    void updateEvent(Integer id, JsonObject event, Handler<Either<String, JsonObject>> handler);

    /**
     * Update reason for each event
     *
     * @param eventBody Event body that can contain list of id's events and its reason_id retrieved
     * @param handler   Function handler returning data
     */
    void changeReasonEvents(JsonObject eventBody, UserInfos user, Handler<Either<String, JsonObject>> handler);

    Future<JsonObject> changeReasonEvents(List<Long> eventsIds, Integer reasonId);

    /**
     * Update regularized for each event
     *
     * @param event   Event body that can contain list of id's events and its reason_id retrieved
     * @param handler Function handler returning data
     */
    void changeRegularizedEvents(JsonObject event, UserInfos user, Handler<Either<String, JsonObject>> handler);


    /**
     * Delete given identifier
     *
     * @param id      event identifier
     * @param handler function handler returning data
     */
    void delete(Integer id, Handler<Either<String, JsonObject>> handler);

    /**
     * List events as strict values based parameters
     *
     * @param structureId Structure identifier
     * @param startDate   Event start date
     * @param endDate     event end date
     * @param eventType   Event type
     * @param userId      User identifiers
     * @param handler     Function handler return data
     */
    void list(String structureId, String startDate, String endDate, List<Integer> eventType, List<String> userId, Handler<Either<String, JsonArray>> handler);

    /**
     * Get count event type group by user
     *
     * @param eventType      Event type
     * @param students       Student list. Contains every students identifiers
     * @param structure      Structure identifier
     * @param justified      Justified events or not ? Can be null if justified event needs to be excluded (DEPRECATED, MUST USE REGULARIZED)
     * @param startAt        Start count at. Minimal number that student is retrieved
     * @param massmailed     Massmailed ? Use by massmailing module. When null, column is excluded
     * @param startDate      Range start date
     * @param endDate        Range end date
     * @param reasonsId      Reasons identifiers. Can be sorted by reasons identifiers
     * @param noReasons      Should retrieve no reasons event
     * @param recoveryMethod recovery method to fetch data used
     * @param regularized    filter regularized (replace justified)
     * @param handler        Function handler returning data
     */
    void getCountEventByStudent(Integer eventType, List<String> students, String structure, Boolean justified,
                                Integer startAt, List<Integer> reasonsId, Boolean massmailed, String startDate,
                                String endDate, boolean noReasons, String recoveryMethod, Boolean regularized,
                                Handler<Either<String, JsonArray>> handler);

    /**
     * Retrieve events by student
     *
     * @param eventType          Event Type list
     * @param students           Student list. Contains every students identifiers
     * @param structure          Structure identifier
     * @param justified          Justified events or not ? Can be null if justified event needs to be excluded
     * @param massmailed         Massmailed ? Use by massmailing module. When null, column is excluded
     * @param startDate          Range start date
     * @param endDate            Range end date
     * @param reasonsId          Reasons identifiers. Can be sorted by reasons identifiers
     * @param noReasons          Should retrieve no reasons event
     * @param recoveryMethodUsed method used to recover events, can be null if method in settings is wanted.
     * @param limit              corresponding to the limit of data rows wanted (optional)   method used to recover events, can be null if method in settings is wanted.
     * @param offset             corresponding to the offset data rows wanted (optional).
     * @param handler            Function handler returning data
     */
    void getEventsByStudent(Integer eventType, List<String> students, String structure, List<Integer> reasonsId, Boolean massmailed,
                            String startDate, String endDate, boolean noReasons, String recoveryMethodUsed, String limit, String offset,
                            Boolean regularized, Handler<Either<String, JsonArray>> handler);

    Future<JsonArray> getEventsByStudent(Boolean canSeeAllStudent, Integer eventType, List<String> students, String structure,
                                         List<Integer> reasonsId, Boolean massmailed, Boolean compliance, String startDate, String endDate,
                                         boolean noReasons, boolean noReasonsLateness, String recoveryMethodUsed, String limit, String offset,
                                         Boolean regularized, Boolean followed);

    void getEventsByStudent(Boolean canSeeAllStudent, Integer eventType, List<String> students, String structure,
                            List<Integer> reasonsId, Boolean massmailed, Boolean compliance, String startDate, String endDate,
                            boolean noReasons, Boolean noReasonsLateness, String recoveryMethodUsed, String limit, String offset,
                            Boolean regularized, Boolean followed, Handler<Either<String, JsonArray>> handler);

    void getEventsByStudent(Integer eventType, List<String> students, String structure, List<Integer> reasonsId,
                            Boolean massmailed, String startDate, String endDate, boolean noReasons, String recoveryMethodUsed,
                            Boolean regularized, Handler<Either<String, JsonArray>> handler);

    /**
     * Retrieve events by student
     *
     * @param eventType          Event Type list
     * @param students           Student list. Contains every students identifiers
     * @param structure          Structure identifier
     * @param massmailed         Massmailed ? Use by massmailing module. When null, column is excluded
     * @param compliance         compliance ? Using this filter will remove absence whose compliance is TRUE if included
     * @param startDate          Range start date
     * @param endDate            Range end date
     * @param reasonsId          Reasons identifiers. Can be sorted by reasons identifiers
     * @param noReasons          Should retrieve no reasons event
     * @param recoveryMethodUsed method used to recover events, can be null if method in settings is wanted.
     * @param handler            Function handler returning data
     */
    void getEventsByStudent(Integer eventType, List<String> students, String structure, List<Integer> reasonsId,
                            Boolean massmailed, Boolean compliance, String startDate, String endDate, boolean noReasons,
                            String recoveryMethodUsed, Boolean regularized, Handler<Either<String, JsonArray>> handler);

    /**
     * get event action
     *
     * @param eventId event identifier
     * @param handler Function handler returning data
     */
    void getActions(String eventId, Handler<Either<String, JsonArray>> handler);

    /**
     * create event action
     *
     * @param actionBody actionBody fetched
     * @param handler    Function handler returning data
     */
    void createAction(JsonObject actionBody, Handler<Either<String, JsonObject>> handler);


    /**
     * Retrieve student absence rate for given period
     *
     * @param student   student identifier
     * @param structure structure identifier
     * @param start     start date
     * @param end       end date
     * @param handler   function handler returning data
     */
    void getAbsenceRate(String student, String structure, String start, String end, Handler<Either<String, JsonObject>> handler);
}

File: presences/src/main/resources/public/ts/services/EventService.ts
import {ng} from 'entcore';
import http, {AxiosResponse} from 'axios';
import {ActionBody, EventResponse, Events, IEventBody, IStudentEventRequest, IStudentEventResponse} from '../models';
import {DateUtils} from '@common/utils';
import {EventAbsenceSummary} from '@presences/models/Event/EventAbsenceSummary';
import {EXPORT_TYPE, ExportType} from "@common/core/enum/export-type.enum";

export interface EventService {
    get(eventRequest: EventRequest): Promise<{ pageCount: number, events: EventResponse[], all: EventResponse[] }>;

    getEventActions(event_id: number): Promise<ActionBody[]>;

    createAction(actionBody: ActionBody): Promise<AxiosResponse>;

    getStudentEvent(studentEventRequest: IStudentEventRequest): Promise<IStudentEventResponse>;

    export(eventRequest: EventRequest, exportType: ExportType): string;

    getAbsentsCounts(structureId: string, startDate: string, endDate: string): Promise<EventAbsenceSummary>;

    createLatenessEvent(eventBody: IEventBody, structureId: string): Promise<AxiosResponse>;

    updateEvent(eventId: number, eventBody: IEventBody): Promise<AxiosResponse>;

    deleteEvent(eventId: number): Promise<AxiosResponse>;
}

export interface EventRequest {
    structureId: string;
    startDate: string;
    endDate: string;
    startTime?: string;
    endTime?: string;
    noReason?: boolean;
    noReasonLateness?: boolean;
    eventType: string;
    listReasonIds?: string;
    regularized?: boolean;
    followed?: boolean;
    notFollowed?: boolean;
    userId?: string;
    userIds?: string[];
    classes?: string;
    classesIds?: string[];
    page?: number;
}

export const eventService: EventService = {
    get: async (eventRequest: EventRequest): Promise<{ pageCount: number, events: EventResponse[], all: EventResponse[] }> => {
        let dateFormat = DateUtils.FORMAT['YEAR-MONTH-DAY'];
        try {
            const structureId: string = `?structureId=${eventRequest.structureId}`;
            const startDate: string = `&startDate=${DateUtils.format(eventRequest.startDate, dateFormat)}`;
            const endDate: string = `&endDate=${DateUtils.format(eventRequest.endDate, dateFormat)}`;
            const startTime: string = eventRequest.startTime ? `&startTime=${eventRequest.startTime}` : '';
            const endTime: string = eventRequest.endTime ? `&endTime=${eventRequest.endTime}` : '';
            const noReason: string = eventRequest.noReason ? `&noReason=${eventRequest.noReason}` : '';
            const noReasonLateness: string = eventRequest.noReasonLateness ? `&noReasonLateness=${eventRequest.noReasonLateness}` : '';
            const eventType: string = eventRequest.eventType ? `&eventType=${eventRequest.eventType}` : '';
            const listReasonIds: string = eventRequest.listReasonIds ? `&reasonIds=${eventRequest.listReasonIds}` : '';
            const userId: string = eventRequest.userId && eventRequest.userId.length > 0 ? `&userId=${eventRequest.userId}` : '';

            let userIds: string = '';
            if (eventRequest.userIds && eventRequest.userIds.length > 0) {
                eventRequest.userIds.forEach(userId => userIds += `&userId=${userId}`);
            }

            const classes: string = eventRequest.classes && eventRequest.classes.length > 0 ? `&classes=${eventRequest.classes}` : '';

            let classesIds: string = '';
            if (eventRequest.classesIds && eventRequest.classesIds.length > 0) {
                eventRequest.classesIds.forEach(classId => classesIds += `&classes=${classId}`);
            }
            const regularized: string = (eventRequest.regularized !== undefined && eventRequest.regularized !== null)
                ? `&regularized=${eventRequest.regularized}` : '';

            const followed: string = (eventRequest.followed != null || eventRequest.notFollowed != null)
                    && (eventRequest.followed === !eventRequest.notFollowed) ? `&followed=${eventRequest.followed}` : '';

            const page: string = eventRequest.page ? `&page=${eventRequest.page}` : '';


            const urlParams: string = `${structureId}${startDate}${endDate}${startTime}${endTime}${noReason}${noReasonLateness}` +
            `${eventType}${listReasonIds}${userId}${userIds}${classes}${classesIds}${regularized}${followed}${page}`;

            const {data}: AxiosResponse = await http.get(`/presences/events${urlParams}`);

            return {
                pageCount: data.page_count,
                events: data.all,
                all: Events.buildEventResponse(data.all, eventRequest.page)
            };
        } catch (err) {
            throw err;
        }
    },
    
    getEventActions: async (eventId: number): Promise<ActionBody[]> => {
        try {
            const {data} = await http.get(`/presences/events/${eventId}/actions`);
            return data;
        } catch (err) {
            throw err;
        }
    },

    createAction: async (actionBody: ActionBody): Promise<AxiosResponse> => {
        return http.post(`/presences/events/actions`, actionBody);
    },

    getStudentEvent: async (studentEventRequest: IStudentEventRequest): Promise<IStudentEventResponse> => {
        try {
            const structure_id: string = `?structure_id=${studentEventRequest.structure_id}`;
            const start_at: string = `&start_at=${studentEventRequest.start_at}`;
            const end_at: string = `&end_at=${studentEventRequest.end_at}`;

            let types: string = '';
            if (studentEventRequest.type) {
                studentEventRequest.type.forEach((type: string) => {
                    types += `&type=${type}`;
                });
            }

            let limit: string = '';
            if (studentEventRequest.limit) {
                limit = `&limit=${studentEventRequest.limit}`;
            }

            let offset: string = '';
            if (studentEventRequest.offset) {
                offset = `&offset=${studentEventRequest.offset}`;
            }

            const urlParams = `${structure_id}${start_at}${end_at}${types}${limit}${offset}`;
            const {data} = await http.get(`/presences/students/${studentEventRequest.student_id}/events${urlParams}`);
            return data
        } catch (err) {
            throw err;
        }
    },

    export(eventRequest: EventRequest, exportType: ExportType): string {
        const dateFormat: string = DateUtils.FORMAT['YEAR-MONTH-DAY'];

        const url: string = `/presences/events/export`;
        const type: string = `?type=${EXPORT_TYPE[exportType]}`;
        const structureId: string = `&structureId=${eventRequest.structureId}`;
        const startDate: string = `&startDate=${DateUtils.format(eventRequest.startDate, dateFormat)}`;
        const endDate: string = `&endDate=${DateUtils.format(eventRequest.endDate, dateFormat)}`;
        const noReason: string = eventRequest.noReason ? `&noReason=${eventRequest.noReason}` : "";
        const eventType: string = `&eventType=${eventRequest.eventType}`;
        const listReasonIds: string = eventRequest.listReasonIds ? `&reasonIds=${eventRequest.listReasonIds}` : "";
        const userId: string = eventRequest.userId.length === 0 ? "" : `&userId=${eventRequest.userId}`;
        const classes: string = eventRequest.classes.length === 0 ? "" : `&classes=${eventRequest.classes}`;
        const regularized: string = (eventRequest.regularized != null) ? `&regularized=${(eventRequest.regularized)}` : "";
        const basedUrl: string = `${url}${type}${structureId}`;

        return `${basedUrl}${startDate}${endDate}${noReason}${eventType}${listReasonIds}${userId}${classes}${regularized}`;
    },


    /**
     * Get counts of current absences.
     *
     * @param structureId       structure id
     * @param startDate         start date filter (optional)
     * @param endDate           end date filter (optional)
     */
    async getAbsentsCounts(structureId: string, startDate: string, endDate: string): Promise<EventAbsenceSummary> {
        try {
            let url: string = `/presences/events/absences/summary`;

            if (structureId) {
                url += `?structureId=${structureId}`;
                url += `&currentDate=${DateUtils.getCurrentDate(DateUtils.FORMAT['YEAR-MONTH-DAY-HOUR-MIN-SEC'])}`;
            }

            if (startDate && endDate) {
                url += `&startAt=${startDate}&endAt=${endDate}`;
            }

            const {data}: AxiosResponse = await http.get(url);
            return data;
        } catch (err) {
            throw err;
        }
    },

    /**
     * Create Lateness Event (special treatment (see API)
     *
     * @param eventBody     eventBody
     * @param structureId   structure identifier
     */
    createLatenessEvent(eventBody: IEventBody, structureId: string): Promise<AxiosResponse> {
        return http.post(`/presences/events/${structureId}/lateness`, eventBody);
    },

    /**
     * Update Event
     *
     * @param eventId       event identifier
     * @param eventBody     eventBody
     */
    updateEvent(eventId: number, eventBody: IEventBody): Promise<AxiosResponse> {
        return http.put(`/presences/events/${eventId}`, eventBody);
    },

    /**
     * Delete Event
     *
     * @param eventId       event identifier
     */
    deleteEvent(eventId: number): Promise<AxiosResponse> {
        return http.delete(`/presences/events/${eventId}`);
    }
};

export const EventService = ng.service('EventService', (): EventService => eventService);
